torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -720,6 +720,9 @@ void ProcessGroupNCCL::WorkNCCL::handleException(
 
 void ProcessGroupNCCL::WorkNCCL::synchronize() {
   synchronizeStream();
+  c10d::unregister_work(
+      c10::intrusive_ptr<
+          ProcessGroupNCCL::WorkNCCL>::unsafe_reclaim_from_nonowning(this));
```

Comment:
seems that both `WorkNCCL::wait()` and `WorkNCCL::synchronize()` are valid "wait" APIs, so putting unregister_work in `synchronize()` since it's also called by `wait()`